### PR TITLE
Add cli commands for users and groups

### PR DESCRIPTION
**Pull request by pfeuffer**
## Proposed changes

Adds cli commands to manage users and groups. 

### Your checklist for this pull request

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

**Contributor**:
- [x] PR is well described and the description can be used as a commit message on squash
- [ ] Related issues linked to PR if existing and labels set
- [x] New code is covered with unit tests
- [ ] New/updated ui components are tested inside the storybook (module ui-components only) 
- [x] [Changelog entry file](https://github.com/scm-manager/changelog#changelog-entry-files) created in `gradle/changelog`
- [ ] Feature has been tested with different permissions
- [ ] Documentation updated (only necessary for new features or changed behaviour)

**Reviewer**:
- [ ] The clean code principles are respected ([CleanCode](https://clean-code-developer.com/virtues/))
- [ ] All new code/logic is implemented on the right spot / "Should this be done here?"
- [ ] UI changes fits into the layout
- [ ] The UI views / components are responsive (mobile views)
- [ ] Correct translations are available

### Checklist for branch merge request (not required for forks)

- [ ] Branch is green/blue on [Jenkins](https://oss.cloudogu.com/jenkins/)
- [ ] Quality Gate passed on [SonarQube](https://sonarcloud.io/organizations/scm-manager/projects)

### CODE REVIEW

File: scm-webapp/src/main/java/sonia/scm/user/cli/UserDeactivateCommand.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import picocli.CommandLine;
+import sonia.scm.cli.ParentCommand;
+import sonia.scm.user.User;
+import sonia.scm.user.UserManager;
+
+import javax.inject.Inject;
+
+@ParentCommand(value = UserCommand.class)
+@CommandLine.Command(name = "deactivate")
+public class UserDeactivateCommand implements Runnable {
```
Comment (eheimbuch) [inline scm-webapp/src/main/java/sonia/scm/user/cli/UserDeactivateCommand.java:36]: public?

File: scm-webapp/src/main/java/sonia/scm/user/cli/UserDeleteCommand.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import sonia.scm.cli.ParentCommand;
+import sonia.scm.user.User;
+import sonia.scm.user.UserManager;
+
+import javax.inject.Inject;
+import java.util.Collections;
+
+@CommandLine.Command(name = "delete", aliases = "rm")
+@ParentCommand(UserCommand.class)
+public class UserDeleteCommand implements Runnable {
```
Comment (eheimbuch) [inline scm-webapp/src/main/java/sonia/scm/user/cli/UserDeleteCommand.java:38]: public?

File: scm-webapp/src/main/java/sonia/scm/user/cli/UserGetCommand.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import picocli.CommandLine;
+import sonia.scm.cli.ParentCommand;
+import sonia.scm.user.User;
+import sonia.scm.user.UserManager;
+
+import javax.inject.Inject;
+
+@ParentCommand(value = UserCommand.class)
+@CommandLine.Command(name = "get")
+public class UserGetCommand implements Runnable {
```
Comment (eheimbuch) [inline scm-webapp/src/main/java/sonia/scm/user/cli/UserGetCommand.java:36]: public?

File: scm-webapp/src/main/java/sonia/scm/user/cli/UserListCommand.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import sonia.scm.cli.TemplateRenderer;
+import sonia.scm.user.UserManager;
+
+import javax.inject.Inject;
+import java.util.Collection;
+import java.util.stream.Collectors;
+
+@ParentCommand(value = UserCommand.class)
+@CommandLine.Command(name = "list", aliases = "ls")
+public class UserListCommand implements Runnable {
```
Comment (eheimbuch) [inline scm-webapp/src/main/java/sonia/scm/user/cli/UserListCommand.java:41]: public?

File: scm-webapp/src/main/java/sonia/scm/user/cli/UserModifyCommand.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import sonia.scm.cli.ParentCommand;
+import sonia.scm.user.User;
+import sonia.scm.user.UserManager;
+
+import javax.inject.Inject;
+import javax.validation.constraints.Email;
+
+@ParentCommand(value = UserCommand.class)
+@CommandLine.Command(name = "modify")
+public class UserModifyCommand implements Runnable {
```
Comment (eheimbuch) [inline scm-webapp/src/main/java/sonia/scm/user/cli/UserModifyCommand.java:39]: public?

File: scm-webapp/src/main/java/sonia/scm/user/cli/UserConvertToInternalCommand.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+  @CommandLine.Mixin
+  private final UserTemplateRenderer templateRenderer;
+  private final UserManager manager;
+
+  @CommandLine.Parameters(index = "0", paramLabel = "<username>", descriptionKey = "scm.user.username")
+  private String username;
+
+  @CommandLine.Parameters(index = "1", paramLabel = "<password>", descriptionKey = "scm.user.password")
+  private String password;
```
Comment (eheimbuch) [inline scm-webapp/src/main/java/sonia/scm/user/cli/UserConvertToInternalCommand.java:47]: Password validation? Should be between 6 and 32 characters.

File: scm-webapp/src/main/resources/sonia/scm/cli/i18n.properties [hunk 7]
```diff
@@ -79,3 +84,102 @@ repoDeletePrompt= If you really want to delete this repository please pass --yes
...
+scm.group.description = Description
+scm.group.members = Members
+scm.group.membersList = Members
+scm.group.external = External
+
+### Group command
+scm.group.usage.description.0 = Parent command for all group related actions
+
+### Get group
+scm.group.get.usage.description.0 = Returns group related information
```
Comment (eheimbuch) [inline scm-webapp/src/main/resources/sonia/scm/cli/i18n.properties:148]:
```suggestion
scm.group.get.usage.description.0 = Returns group-related information
```

File: scm-webapp/src/main/resources/sonia/scm/cli/i18n.properties [hunk 8]
```diff
@@ -79,3 +84,102 @@ repoDeletePrompt= If you really want to delete this repository please pass --yes
...
+scmUserErrorPassword = Password is required for internal users
+scmUserErrorExternalActivate = External users cannot be activated
+scmUserErrorExternalDeactivate = External users cannot be deactivated
+scmUserErrorNotFound = Could not find user
+
+### User command
+scm.user.usage.description.0 = Parent command for all user-related actions
+
+### Get user
+scm.user.get.usage.description.0 = Returns user related information
```
Comment (eheimbuch) [inline scm-webapp/src/main/resources/sonia/scm/cli/i18n.properties:106]:
```suggestion
scm.user.get.usage.description.0 = Returns user-related information
```

File: scm-webapp/src/main/resources/sonia/scm/cli/i18n_de.properties [hunk 9]
```diff
@@ -79,3 +84,102 @@ repoDeletePrompt= Wenn dieses Repository endg
...
+scm.user.active = Aktiv
+
+### User error
+scmUserErrorPassword = F�r interne Benutzer muss ein Passwort gesetzt werden
+scmUserErrorExternalActivate = Externe Benutzer k�nnen nicht aktiviert werden
+scmUserErrorExternalDeactivate = Externe Benutzer k�nnen nicht deaktiviert werden
+scmUserErrorNotFound= Benutzer konnte nicht gefunden werden
+
+### User command
+scm.user.usage.description.0 = Gruppen Befehl f�r alle User Befehle
```
Comment (eheimbuch) [inline scm-webapp/src/main/resources/sonia/scm/cli/i18n_de.properties:103]:
```suggestion
scm.user.usage.description.0 = Gruppen Befehl für alle Benutzer Befehle
```

File: scm-webapp/src/main/resources/sonia/scm/cli/i18n_de.properties [hunk 10]
```diff
@@ -79,3 +84,102 @@ repoDeletePrompt= Wenn dieses Repository endg
...
+
+### Group
+scm.group.name = Name
+scm.group.description = Beschreibung
+scm.group.members = Mitglieder
+scm.group.membersList = Mitglieder
+scm.group.external = Extern
+
+### Group command
+scm.group.usage.description.0 = Hauptbefehl f�r alle Gruppen Befehle
```
Comment (eheimbuch) [inline scm-webapp/src/main/resources/sonia/scm/cli/i18n_de.properties:145]:
```suggestion
scm.group.usage.description.0 = Gruppenbefehl für alle \"Gruppen\" Befehle
```

Maybe like this to make it more consistent?

File: scm-webapp/src/main/java/sonia/scm/user/cli/UserModifyCommand.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+  @CommandLine.Parameters(index = "0", paramLabel = "<username>", descriptionKey = "scm.user.username")
+  private String username;
+
+  @CommandLine.Option(names = {"--displayname", "-d"}, descriptionKey = "scm.user.displayName")
+  private String displayName;
+
+  @Email
+  @CommandLine.Option(names = {"--email", "-e"}, descriptionKey = "scm.user.create.email")
+  private String email;
+
```
Comment (eheimbuch) [inline scm-webapp/src/main/java/sonia/scm/user/cli/UserModifyCommand.java:56]: Can I not change passwords with the modify command?

**Comment by eheimbuch** [comment]
Sonar found some duplicate code.

**Review by eheimbuch — CHANGES_REQUESTED** [review]

File: scm-webapp/src/main/resources/sonia/scm/cli/i18n_de.properties (hunk 10 above)
Comment (matthiasthieroff) [reply scm-webapp/src/main/resources/sonia/scm/cli/i18n_de.properties:145]: I don't like neither `Gruppenbefehl` nor "Hauptbefehl". Even "Parent command" sounds a bit weird, but we couldn't find a better way to express what the command does.

File: scm-webapp/src/main/java/sonia/scm/group/cli/GroupCommandBeanMapper.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
+import sonia.scm.group.Group;
+
+import java.time.Instant;
+import java.time.format.DateTimeFormatter;
+
+@Mapper
+interface GroupCommandBeanMapper {
```
[the remark this answers is not in the document]
Comment (matthiasthieroff) [reply scm-webapp/src/main/java/sonia/scm/group/cli/GroupCommandBeanMapper.java:37]: The `GroupCommandBeanMapper` and `UserCommandBeanMapper` are implicitly tested by all tests which test the command output. An instance of the mapper implementation is created in the `TemplateTestRenderer` class.

**Pull request merged**